napari/components/dims.py

```diff
@@ -408,9 +585,62 @@ def _roll(self):
         order[nsteps > 1] = np.roll(order[nsteps > 1], 1)
         self.order = order.tolist()
 
+    def _sanitize_input(self, axis, value, value_is_sequence=False):
+        """
+        Ensure that axis is and value are the same length, that axis are not
+        out of bounds, and corerces to lists for easier processing.
+        """
+        if isinstance(axis, Integral):
+            if (
+                isinstance(value, Sequence)
+                and not isinstance(value, str)
+                and not value_is_sequence
+            ):
+                raise ValueError(
+                    trans._('cannot set multiple values to a single axis')
+                )
+            axis = [axis]
+            value = [value]
+        else:
+            axis = list(axis)
+            value = list(value)
+
+        if len(axis) != len(value):
+            raise ValueError(
+                trans._("axis and value sequences must have equal length")
+            )
+
+        for ax in axis:
+            ensure_axis_in_bounds(ax, self.ndim)
+        return axis, value
+
+
+def ensure_ndim(value: Tuple, ndim: int, default: Tuple):
```

Comment:
Love this as a way to make the root validator easier to read.

```suggestion
def ensure_ndim(value: Tuple, ndim: int, default: Tuple) -> Tuple:
```

Reply:
If you wanted to pull this change out of this PR, I'd be down to review and approve it separately. Won't make a huge difference to the main size/complexity of this PR, but every little helps?

Reply:
Makes sense, I'll see if I can split out!